### ISSUE TITLE
[QUESTION] Capturing Best Iterations for Final Training on Full Dataset

### ISSUE THREAD

**Issue by philipawalkerdatascience** [unit8co/darts#3033]
Hello, I am using Catboost (Regression) within Darts, and have a question regarding best iterations.
With my model, I am using:
- Multi-series (Global Model), 40 series
- Output Chunk 12
- Forecast 12
- Multi-Models = True

I then created a model and fitted it to a train + validation set + early stopping, which found the best iterations, which range from 16 - 200 iterations. I wanted to use these iterations for a final training which combines the entire train/val/test dataset before prediction. However, I am unsure of how to use the captured best iterations for this final training. Is there any way to achieve this? If it is not possible, is there any suggested approach within the Darts framework for using optimal iterations during a final training on the full dataset?

Also, I apologize if this topic has already been covered, I checked the search, but didn't see any similar topics.

Thank you for your time and help!

**Comment by dennisbader** [comment]
Hi @philipawalkerdatascience and sorry for the late response, we were on holidays last week.

You can pass all CatBoostRegressor creation parameters as `kwargs` to the darts model constructor (see below).

```py
from darts.datasets import AusBeerDataset
from darts.models import CatBoostModel

series = AusBeerDataset().load()
model = CatBoostModel(lags=12, output_chunk_length=12, iterations=10)
model.fit(series)
preds = model.predict(n=12)
```

**Comment by philipawalkerdatascience** [comment]
Hi @dennisbader ,

No problem regarding the holidays, and thank you very much for your response.
The difficulty I am having (if I understand correctly), is that when 'multi_models=True', Darts internally creates 12 separate models to predict my forecast_horizon of 12. Each of these models has its own best iteration, as seen here:
```

62:	learn: 0.1686484	test: 0.3505987	best: 0.3502633 (54)	total: 1.84s	remaining: 56.6s
63:	learn: 0.1683403	test: 0.3505807	best: 0.3502633 (54)	total: 1.87s	remaining: 56.6s
64:	learn: 0.1681066	test: 0.3502844	best: 0.3502633 (54)	total: 1.9s	remaining: 56.6s
Stopped by overfitting detector  (10 iterations wait)

bestTest = 0.3502633284
bestIteration = 54

Shrink model to first 55 iterations.
0:	learn: 0.2084955	test: 0.2185743	best: 0.2185743 (0)	total: 23.8ms	remaining: 47.7s
1:	learn: 0.2072178	test: 0.2183550	best: 0.2183550 (1)	total: 55.8ms	remaining: 55.7s
2:	learn: 0.2059817	test: 0.2182541	best: 0.2182541 (2)	total: 88ms	remaining: 58.6s
3:	learn: 0.2050330	test: 0.2182541	best: 0.2182541 (2)	total: 118ms	remaining: 58.9s
```

In my case, the best iterations varied greatly, from 16 iterations being the best for one of these internal models, to 200 for another. Is there any way to replicate these best iterations in a final training set (no validation set) for each of the 12 internal models?

**Comment by dennisbader** [comment]
Hi @philipawalkerdatascience. Unfortunately, setting submodel-specific kwargs is not supported (at the moment at least).
I see two options for you:

### 1. Fit a single multi-output Catboost model on the entire horizon

Using `loss_function="MultiRMSE"` will enable native multioutput support for Catboost and you only have to deal with one set of early stopping / iterations.

```py
from darts.datasets import AusBeerDataset
from darts.models import CatBoostModel

series = AusBeerDataset().load()
train, val = series[:-12], series[-24:]

# first model, find the best iterations
model = CatBoostModel(
    lags=12,
    output_chunk_length=12,
    early_stopping_rounds=5,
    loss_function="MultiRMSE",
    logging_level="Verbose",
)
model.fit(train, val_series=val)

# second model, use those on the complete dataset
model = CatBoostModel(
    lags=12,
    output_chunk_length=12,
    loss_function="MultiRMSE",
    logging_level="Verbose",
    iterations=29,
)
model.fit(series)
preds = model.predict(n=12)
```

This only outputs early stopping logs once:

```
Stopped by overfitting detector  (5 iterations wait)

bestTest = 96.27566497
bestIteration = 28

Shrink model to first 29 iterations.
```

### 2. Fit a dedicated CatBoostRegressor per output step

You can fit a dedicated model per output step leveraging the `output_chunk_shift` parameter to shift the first predicted step of each submodel into the future. Then you can concatenate the predictions of all submodels along the time axis.

```py
from darts import concatenate
from darts.datasets import AusBeerDataset
from darts.models import CatBoostModel

series = AusBeerDataset().load()
train, val = series[:-12], series[-24:]

output_chunk_length = 12
forecast = []
for output_chunk_shift in range(output_chunk_length):
    # find best iterations per submodel
    model = CatBoostModel(
        lags=12,
        output_chunk_length=1,  # each model predicts 1 step
        output_chunk_shift=output_chunk_shift,  # model specific output shift
        early_stopping_rounds=5,
        logging_level="Verbose",
    )
    model.fit(train, val_series=val)

    # second model, use it on the complete dataset
    model = CatBoostModel(
        lags=12,
        output_chunk_length=1,
        output_chunk_shift=output_chunk_shift,
        iterations=model.model.best_iteration_,  # use the best iterations
        logging_level="Verbose",
    )
    model.fit(series)
    # only predict single step per model (it's automatically shifted)
    preds = model.predict(n=1)
    forecast.append(preds)

# concatenate all along the time axis
forecast = concatenate(forecast)
```

You can of course pack this logic into a `ForecastingModel` a `CatBoostModel` subclass to use it with the general Darts API.

**Comment by philipawalkerdatascience** [comment]
Thank you so much, this is perfect!

**Comment by dennisbader** [comment]
Glad it helped :) closing this issue then.
